File: Source/Aladdin.CAPI.STB/CAPI/STB/STB34101/KeyFactory.cs
using System; 
using System.IO; 

namespace Aladdin.CAPI.STB.STB34101
{
    ////////////////////////////////////////////////////////////////////////////////
    // Фабрика кодирования кючей
    ////////////////////////////////////////////////////////////////////////////////
    public class KeyFactory : CAPI.KeyFactory
    {
        // способ кодирования чисел
        private const Math.Endian Endian = Math.Endian.LittleEndian; 

        // конструктор
        public KeyFactory(string keyOID) { this.keyOID = keyOID; }

        // идентификаторы открытых ключей
        public override string KeyOID { get { return keyOID; }} private string keyOID; 
    
	    // способ использования ключа
	    public override KeyUsage GetKeyUsage() 
        { 
            // указать способ использования ключа
            return KeyUsage.DigitalSignature | KeyUsage.CertificateSignature | 
                   KeyUsage.CrlSignature     | KeyUsage.NonRepudiation       | 
                   KeyUsage.KeyEncipherment; 
        }
        // закодировать параметры
        public override ASN1.IEncodable EncodeParameters(CAPI.IParameters parameters) 
        {
            // при указании идентификатора 
            if (parameters is INamedParameters)
            {
                // закодировать идентификатор параметров
                return new ASN1.ObjectIdentifier(((INamedParameters)parameters).Oid); 
            }
            else { 
                // преобразовать тип параметров
                IParameters stbParameters = (IParameters)parameters; 
            
                // проверить корректность данных
                if (stbParameters.Generator.X.Signum != 0) throw new NotSupportedException(); 

                // получить параметры эллиптической кривой
                EC.CurveFp ec = stbParameters.Curve; EC.FieldFp field = ec.Field;

                // указать идентификатор типа поля
                ASN1.ObjectIdentifier fieldType = new ASN1.ObjectIdentifier(
                    ASN1.STB.OID.stb34101_bign_primeField
                ); 
                // закодировать параметры поля
                ASN1.STB.FieldID fieldID = new ASN1.STB.FieldID(
                    fieldType, new ASN1.Integer(field.P)
                ); 
                // определить размер координат
                int l = stbParameters.Order.BitLength / 2;
            
                // закодировать параметры a и b кривой
                byte[] A = Math.Convert.FromBigInteger(ec.A, Endian, l / 4);
                byte[] B = Math.Convert.FromBigInteger(ec.B, Endian, l / 4);
            
                // закодировать точку на эллиптической кривой
                byte[] GY = Math.Convert.FromBigInteger(stbParameters.Generator.Y, Endian, l / 4);
            
                // закодировать параметры эллиптичесой кривой
                ASN1.STB.Curve curve = new ASN1.STB.Curve(new ASN1.OctetString(A), 
                    new ASN1.OctetString(B), new ASN1.BitString(ec.Seed)
                ); 
                // закодировать параметры в целом
                return new ASN1.STB.ECParameters(new ASN1.Integer(1), fieldID, curve, 
                    new ASN1.OctetString(GY), new ASN1.Integer(stbParameters.Order), null
                ); 
            }
        }
        // раскодировать параметры
        public override CAPI.IParameters DecodeParameters(ASN1.IEncodable encoded)
        {
            // раскодировать параметры
            encoded = new ASN1.STB.DomainParameters().Decode(encoded); 
        
            // проверить указание параметров
            if (encoded is ASN1.Null) throw new NotSupportedException(); 
        
            // при указании идентификатора
            if (encoded is ASN1.ObjectIdentifier)
            {
                // раскодировать идентификатор параметров
                string oid = ((ASN1.ObjectIdentifier)encoded).Value; 
            
                // раскодировать набор параметров
                return DecodeParameters(oid, ASN1.STB.ECParameters.Parameters(oid)); 
            }
            // раскодировать параметры
            else return DecodeParameters(null, (ASN1.STB.ECParameters)encoded); 
        }
        // раскодировать параметры
        public IParameters DecodeParameters(string oid, ASN1.STB.ECParameters parameters) 
        {
            // проверить корректность параметров
            if (parameters.Cofactor.Value.IntValue != 1) throw new InvalidDataException(); 
        
            // раскодировать модуль простого поля
            Math.BigInteger modulus = parameters.FieldID.Parameters.Value;
        
            // извлечь параметры эллиптической кривой
            ASN1.STB.Curve curve = parameters.Curve; byte[] seed = curve.Seed.Value; 
        
            // раскодировать коэффициенты a и b
            Math.BigInteger a = Math.Convert.ToBigInteger(curve.A.Value, Endian); 
            Math.BigInteger b = Math.Convert.ToBigInteger(curve.B.Value, Endian); 
        
            // раскодировать координаты базовой точки
            Math.BigInteger gy = Math.Convert.ToBigInteger(parameters.Base.Value, Endian); 
        
            // создать базовую точку
            EC.Point g = new EC.Point(Math.BigInteger.Zero, gy);
            try {         
                // создать эллиптическую кривую
                EC.CurveFp ec = new EC.CurveFp(modulus, a, b, seed); 
        
                // раскодировать порядок базовой точки
                Math.BigInteger q = parameters.Order.Value; if (oid == null)
                {
                    // вернуть параметры алгоритма
                    return new Parameters(ec, g, q); 
                }
                // вернуть параметры алгоритма
                else return new NamedParameters(oid, ec, g, q); 
            }
            // при ошибке изменить тип исключения
            catch (ArgumentException e) { throw new InvalidDataException(e.Message, e); }
        }
        // закодировать открытый ключ
        public override ASN1.ISO.PKIX.SubjectPublicKeyInfo EncodePublicKey(CAPI.IPublicKey publicKey) 
        {
            // выполнить преобразование типа
            IParameters parameters = (IParameters)publicKey.Parameters; 

            // выполнить преобразование типа
            IPublicKey ecPublicKey = (IPublicKey)publicKey; 
        
		    // закодировать параметры ключа
		    ASN1.ISO.AlgorithmIdentifier algorithm = new ASN1.ISO.AlgorithmIdentifier(
			    new ASN1.ObjectIdentifier(keyOID), EncodeParameters(parameters)
		    ); 
            // определить параметр l
            int l = parameters.Order.BitLength / 2; 
        
            // закодировать координаты
            byte[] QX = Math.Convert.FromBigInteger(ecPublicKey.Q.X, Endian, l / 4);
            byte[] QY = Math.Convert.FromBigInteger(ecPublicKey.Q.Y, Endian, l / 4);
        
            // объединить координаты
            ASN1.BitString encoded = new ASN1.BitString(Arrays.Concat(QX, QY));

            // вернуть закодированное представление
            return new ASN1.ISO.PKIX.SubjectPublicKeyInfo(algorithm, encoded); 
        }
        // раскодировать открытый ключ
        public override CAPI.IPublicKey DecodePublicKey(ASN1.ISO.PKIX.SubjectPublicKeyInfo encoded) 
        {
            // раскодировать параметры
            IParameters parameters = (IParameters)DecodeParameters(encoded.Algorithm.Parameters); 

            // определить параметр l
            int l = parameters.Order.BitLength / 2; byte[] xy = encoded.SubjectPublicKey.Value; 
        
            // проверить корректность размера
            if (xy.Length != l / 2) throw new InvalidDataException(); 
        
            // раскодировать координаты
            Math.BigInteger QX = Math.Convert.ToBigInteger(xy,     0, l / 4, Endian); 
            Math.BigInteger QY = Math.Convert.ToBigInteger(xy, l / 4, l / 4, Endian); 
        
            // создать точку на эллиптической кривой
            EC.Point q = new EC.Point(QX, QY); 

            // вернуть открытый ключ
            return new PublicKey(this, parameters, q); 
        }
        // закодировать личный ключ
        public override ASN1.ISO.PKCS.PKCS8.PrivateKeyInfo EncodePrivateKey(
            CAPI.IPrivateKey privateKey, ASN1.ISO.Attributes attributes) 
        {
            // выполнить преобразование типа
            IParameters parameters = (IParameters)privateKey.Parameters; 

            // выполнить преобразование типа
            IPrivateKey ecPrivateKey = (IPrivateKey)privateKey; 
        
		    // закодировать параметры ключа
		    ASN1.ISO.AlgorithmIdentifier algorithm = new ASN1.ISO.AlgorithmIdentifier(
			    new ASN1.ObjectIdentifier(keyOID), EncodeParameters(parameters)
		    ); 
            // определить параметр l
            int l = parameters.Order.BitLength / 2; 
        
            // закодировать значение личного ключа
            ASN1.OctetString encoded = new ASN1.OctetString(
                Math.Convert.FromBigInteger(ecPrivateKey.D, Endian, l / 4)
            );
            // вернуть закодированное представление
            return new ASN1.ISO.PKCS.PKCS8.PrivateKeyInfo(
                new ASN1.Integer(0), algorithm, encoded, attributes
            ); 
        }
        // раскодировать личный ключ
        public override CAPI.IPrivateKey DecodePrivateKey(CAPI.Factory factory,
            ASN1.ISO.PKCS.PKCS8.PrivateKeyInfo encoded) 
        {
            // раскодировать параметры
            IParameters parameters = (IParameters)DecodeParameters(
                encoded.PrivateKeyAlgorithm.Parameters
            ); 
            // определить параметр l
            int l = parameters.Order.BitLength / 2; byte[] encodedD = encoded.PrivateKey.Value; 
        
            // проверить корректность размера
            if (encodedD.Length != l / 4) throw new InvalidDataException(); 
        
            // раскодировать значение личного ключа
            Math.BigInteger d = Math.Convert.ToBigInteger(encodedD, Endian); 
        
            // вернуть личный ключ
            return new PrivateKey(factory, null, keyOID, parameters, d); 
        }
        // закодировать пару ключей
        public override ASN1.ISO.PKCS.PKCS8.PrivateKeyInfo EncodeKeyPair(
            CAPI.IPrivateKey privateKey, CAPI.IPublicKey publicKey, 
            ASN1.ISO.Attributes attributes)
        {
            // закодировать личный ключ
            return EncodePrivateKey(privateKey, attributes); 
        }
	    // раскодировать пару ключей
        public override KeyPair DecodeKeyPair(CAPI.Factory factory,  
            ASN1.ISO.PKCS.PKCS8.PrivateKeyInfo encoded) 
        {
            // раскодировать личный ключ
            using (IPrivateKey privateKey = (IPrivateKey)DecodePrivateKey(factory, encoded))
            {  
                // преобразовать тип параметров
                IParameters ecParameters = (IParameters)privateKey.Parameters; 

		        // вычислить открытый ключ
		        EC.Point Q = ecParameters.Curve.Multiply(ecParameters.Generator, privateKey.D); 

                // создать объект открытого ключа 
                IPublicKey publicKey = new PublicKey(this, ecParameters, Q);

                // вернуть пару ключей
                return new KeyPair(publicKey, privateKey, null); 
            }
        } 
    }
}

File: Source/Aladdin.CAPI.STB.Java/src/aladdin/capi/stb/stb34101/KeyFactory.java
package aladdin.capi.stb.stb34101;
import aladdin.math.*;
import aladdin.asn1.*;
import aladdin.asn1.Integer;
import aladdin.asn1.iso.*; 
import aladdin.asn1.iso.pkix.*; 
import aladdin.asn1.iso.pkcs.pkcs8.*; 
import aladdin.asn1.stb.*;
import aladdin.capi.*; 
import aladdin.capi.ec.*; 
import aladdin.util.*;
import java.security.spec.*; 
import java.math.*;
import java.io.*;

////////////////////////////////////////////////////////////////////////////////
// Фабрика кодирования кючей
////////////////////////////////////////////////////////////////////////////////
public class KeyFactory extends aladdin.capi.KeyFactory
{
    // способ кодирования чисел
    private static final Endian ENDIAN = Endian.LITTLE_ENDIAN; 

    // конструктор
    public KeyFactory(String keyOID) { this.keyOID = keyOID; }

    // идентификаторы открытых ключей
    @Override public String keyOID() { return keyOID; } private final String keyOID; 
    
	// способ использования ключа
	@Override public KeyUsage getKeyUsage() 
    { 
        // указать способ использования ключа
        return new KeyUsage(KeyUsage.DIGITAL_SIGNATURE | 
            KeyUsage.CERTIFICATE_SIGNATURE | KeyUsage.CRL_SIGNATURE | 
            KeyUsage.NON_REPUDIATION       | KeyUsage.KEY_ENCIPHERMENT
        );
    }
    // закодировать параметры
    @Override public IEncodable encodeParameters(aladdin.capi.IParameters parameters) 
    {
        // при указании идентификатора 
        if (parameters instanceof NamedParameters)
        {
            // закодировать идентификатор параметров
            return new ObjectIdentifier(((NamedParameters)parameters).oid()); 
        }
        else { 
            // преобразовать тип параметров
            IParameters stbParameters = (IParameters)parameters; 
            
            // проверить корректность данных
            if (stbParameters.getGenerator().getAffineX().signum() != 0) 
            {
                // при ошибке выбросить исключение
                throw new UnsupportedOperationException(); 
            }
            // получить параметры эллиптической кривой
            CurveFp ec = stbParameters.getCurve(); ECFieldFp field = ec.getField();
            
            // указать идентификатор типа поля
            ObjectIdentifier fieldType = new ObjectIdentifier(
                aladdin.asn1.stb.OID.STB34101_BIGN_PRIMEFIELD
            ); 
            // закодировать параметры поля
            FieldID fieldID = new FieldID(fieldType, new Integer(field.getP())); 
            
            // определить размер координат
            int l = stbParameters.getOrder().bitLength() / 2;

            // закодировать параметры a и b кривой
            byte[] A  = Convert.fromBigInteger(ec.getA(), ENDIAN, l / 4);
            byte[] B  = Convert.fromBigInteger(ec.getB(), ENDIAN, l / 4);
            
            // закодировать точку на эллиптической кривой
            byte[] GY = Convert.fromBigInteger(stbParameters.getGenerator().getAffineY(), ENDIAN, l / 4);
            
            // закодировать параметры эллиптичесой кривой
            aladdin.asn1.stb.Curve curve = new aladdin.asn1.stb.Curve(
                new OctetString(A), new OctetString(B), new BitString(ec.getSeed())
            ); 
            // закодировать параметры в целом
            return new ECParameters(new Integer(1), fieldID, curve, 
                new OctetString(GY), new Integer(stbParameters.getOrder()), null
            ); 
        }
    }
    // раскодировать параметры
    @Override public IParameters decodeParameters(IEncodable encoded) throws IOException 
    {
        // раскодировать параметры
        encoded = new DomainParameters().decode(encoded); 
        
        // проверить указание параметров
        if (encoded instanceof Null) throw new UnsupportedOperationException(); 
        
        // при указании идентификатора
        if (encoded instanceof ObjectIdentifier)
        {
            // раскодировать идентификатор параметров
            String oid = ((ObjectIdentifier)encoded).value(); 
            
            // раскодировать набор параметров
            return decodeParameters(oid, ECParameters.parameters(oid)); 
        }
        // раскодировать параметры
        else return decodeParameters(null, (ECParameters)encoded); 
    }
    // раскодировать параметры
    public IParameters decodeParameters(String oid, ECParameters parameters) throws IOException 
    {
        // проверить корректность параметров
        if (parameters.cofactor().value().intValue() != 1) throw new IOException(); 
        
        // раскодировать модуль простого поля
        BigInteger modulus = parameters.fieldID().parameters().value();
        
        // извлечь параметры эллиптической кривой
        aladdin.asn1.stb.Curve curve = parameters.curve(); byte[] seed = curve.seed().value(); 
        
        // раскодировать коэффициенты a и b
        BigInteger a = Convert.toBigInteger(curve.a().value(), ENDIAN); 
        BigInteger b = Convert.toBigInteger(curve.b().value(), ENDIAN); 
        
        // создать эллиптическую кривую
        CurveFp ec = new CurveFp(modulus, a, b, seed); 
        
        // раскодировать координаты базовой точки
        BigInteger gy = Convert.toBigInteger(parameters.base().value(), ENDIAN); 
        
        // создать базовую точку
        ECPoint g = new ECPoint(BigInteger.ZERO, gy); 
        
        // раскодировать порядок базовой точки
        BigInteger q = parameters.order().value(); if (oid == null)
        {
            // вернуть параметры алгоритма
            return new Parameters(ec, g, q); 
        }
        // вернуть параметры алгоритма
        else return new NamedParameters(oid, ec, g, q); 
    }
    // закодировать открытый ключ
    @Override public SubjectPublicKeyInfo encodePublicKey(aladdin.capi.IPublicKey publicKey) 
    {
        // выполнить преобразование типа
        IParameters parameters = (IParameters)publicKey.parameters(); 
        
        // выполнить преобразование типа
        IPublicKey ecPublicKey = (IPublicKey)publicKey; 
        
		// закодировать параметры ключа
		AlgorithmIdentifier algorithm = new AlgorithmIdentifier(
			new ObjectIdentifier(keyOID), encodeParameters(parameters)
		); 
        // определить параметр l
        int l = parameters.getOrder().bitLength() / 2; 
        
        // закодировать координаты
        byte[] QX = Convert.fromBigInteger(ecPublicKey.getW().getAffineX(), ENDIAN, l / 4);
        byte[] QY = Convert.fromBigInteger(ecPublicKey.getW().getAffineY(), ENDIAN, l / 4);
        
        // закодировать значение ключа
        BitString encoded = new BitString(Array.concat(QX, QY));

        // вернуть закодированное представление
        return new SubjectPublicKeyInfo(algorithm, encoded); 
    }
    // раскодировать открытый ключ
    @Override public aladdin.capi.IPublicKey decodePublicKey(
        SubjectPublicKeyInfo encoded) throws IOException 
    {
        // раскодировать параметры
        IParameters parameters = decodeParameters(encoded.algorithm().parameters()); 
        
        // определить параметр l
        int l = parameters.getOrder().bitLength() / 2; byte[] xy = encoded.subjectPublicKey().value(); 
        
        // проверить корректность размера
        if (xy.length != l / 2) throw new IOException(); 
        
        // раскодировать координаты
        BigInteger QX = Convert.toBigInteger(xy,     0, l / 4, ENDIAN); 
        BigInteger QY = Convert.toBigInteger(xy, l / 4, l / 4, ENDIAN); 
        
        // раскодировать значение открытого ключа
        ECPoint q = new ECPoint(QX, QY); 
        
        // вернуть открытый ключ
        return new PublicKey(this, parameters, q); 
    }
    // закодировать личный ключ
    @Override public PrivateKeyInfo encodePrivateKey(
        aladdin.capi.IPrivateKey privateKey, Attributes attributes) 
    {
        // выполнить преобразование типа
        IParameters parameters = (IParameters)privateKey.parameters();
        
        // выполнить преобразование типа
        IPrivateKey ecPrivateKey = (IPrivateKey)privateKey; 
        
		// закодировать параметры ключа
		AlgorithmIdentifier algorithm = new AlgorithmIdentifier(
			new ObjectIdentifier(keyOID), encodeParameters(parameters)
		); 
        // определить параметр l
        int l = parameters.getOrder().bitLength() / 2; 
        
        // закодировать значение личного ключа
        OctetString encoded = new OctetString(
            Convert.fromBigInteger(ecPrivateKey.getS(), ENDIAN, l / 4)
        );
        // вернуть закодированное представление
        return new PrivateKeyInfo(new Integer(0), algorithm, encoded, attributes); 
    }
    // раскодировать личный ключ
    @Override public aladdin.capi.IPrivateKey decodePrivateKey(
        Factory factory, PrivateKeyInfo encoded) throws IOException 
    {
        // раскодировать параметры
        IParameters parameters = decodeParameters(
            encoded.privateKeyAlgorithm().parameters()
        ); 
        // определить параметр l
        int l = parameters.getOrder().bitLength() / 2; 
        
        // звлечь закодированное представление
        byte[] encodedD = encoded.privateKey().value(); 
        
        // проверить корректность размера
        if (encodedD.length != l / 4) throw new IOException(); 
        
        // раскодировать значение личного ключа
        BigInteger d = Convert.toBigInteger(encodedD, ENDIAN); 
        
        // вернуть личный ключ
        return new PrivateKey(factory, null, keyOID, parameters, d); 
    }
    // закодировать пару ключей
    @Override public final PrivateKeyInfo encodeKeyPair(
        aladdin.capi.IPrivateKey privateKey, aladdin.capi.IPublicKey publicKey, 
        Attributes attributes) throws IOException
    {
        // закодировать личный ключ
        return encodePrivateKey(privateKey, attributes); 
    }
	// раскодировать пару ключей
    @Override public KeyPair decodeKeyPair(Factory factory, 
        PrivateKeyInfo encoded) throws IOException
    {
        // раскодировать личный ключ
        try (IPrivateKey privateKey = (IPrivateKey)decodePrivateKey(factory, encoded))
        {  
            // преобразовать тип параметров
            IParameters ecParameters = (IParameters)privateKey.parameters(); 

            // вычислить открытый ключ
		    ECPoint Q = ecParameters.getCurve().multiply(
                ecParameters.getGenerator(), privateKey.getS()
            ); 
            // создать объект открытого ключа 
            IPublicKey publicKey = new PublicKey(this, ecParameters, Q);

            // вернуть пару ключей
            return new KeyPair(publicKey, privateKey, null); 
        }
    } 
    // создать параметры
    @Override public aladdin.capi.IParameters createParameters(
        AlgorithmParameterSpec paramSpec) throws InvalidParameterSpecException
    {
        // создать параметры
        return Parameters.getInstance(paramSpec); 
    }
    // создать открытый ключ
    @Override public aladdin.capi.IPublicKey createPublicKey(KeySpec keySpec) 
        throws InvalidKeySpecException, IOException
    {
        // проверить требуемый тип данных
        if (keySpec instanceof ECPublicKeySpec)
        {
            // выполнить преобразование типа
            ECPublicKeySpec ecKeySpec = (ECPublicKeySpec)keySpec; 
            
            // получить параметры 
            ECParameterSpec ecParamSpec = ecKeySpec.getParams(); 
            
            // проверить корректность параметров
            if (ecParamSpec.getCofactor() != 1) throw new InvalidKeySpecException(); 
            
            // в зависимости от типа параметров
            if (ecParamSpec instanceof IParameters)
            {            
                // выполнить преобразование типа
                IParameters parameters = (IParameters)ecParamSpec; 
                
                // создать открытый ключ
                return new PublicKey(this, parameters, ecKeySpec.getW()); 
            }
            try { 
                // преобразовать тип кривой
                CurveFp curve = CurveFp.convert(ecParamSpec.getCurve());
        
                // создать параметры ключа
                IParameters parameters = new Parameters(curve, 
                    ecParamSpec.getGenerator(), ecParamSpec.getOrder()
                ); 
                // создать открытый ключ
                return new PublicKey(this, parameters, ecKeySpec.getW()); 
            }
            // при возникновении ошибки
            catch (IllegalArgumentException e) 
            { 
                // изменить тип исключения
                throw new InvalidKeySpecException(e.getMessage()); 
            }
        }
        // вызвать базовую функцию
        return super.createPublicKey(keySpec); 
    }
    // извлечь данные открытого ключа
    @Override public KeySpec getPublicKeySpec(
        aladdin.capi.IPublicKey publicKey, Class<? extends KeySpec> specType)
        throws InvalidKeySpecException
    {
        // выполнить преобразование типа
        IParameters parameters = (IParameters)publicKey.parameters(); 
        
        // выполнить преобразование типа
        IPublicKey ecPublicKey = (IPublicKey)publicKey; 
        
        // в зависимости от типа данных
        if (specType.isAssignableFrom(ECPublicKeySpec.class))
        {
            // вернуть данные ключа
            return new ECPublicKeySpec(ecPublicKey.getW(), 
                Parameters.convert(parameters)
            ); 
        }
        // вызвать базовую функцию
        return super.getPublicKeySpec(publicKey, specType); 
    }
    // создать личный ключ
    @Override public aladdin.capi.IPrivateKey createPrivateKey(
        Factory factory, KeySpec keySpec) 
        throws InvalidKeySpecException, IOException
    {
        // проверить требуемый тип данных
        if (keySpec instanceof ECPrivateKeySpec)
        {
            // выполнить преобразование типа
            ECPrivateKeySpec ecKeySpec = (ECPrivateKeySpec)keySpec; 
            
            // получить параметры 
            ECParameterSpec ecParamSpec = ecKeySpec.getParams(); 
            
            // проверить корректность параметров
            if (ecParamSpec.getCofactor() != 1) throw new InvalidKeySpecException(); 
            
            // в зависимости от типа параметров
            if (ecParamSpec instanceof IParameters)
            {            
                // выполнить преобразование типа
                IParameters parameters = (IParameters)ecParamSpec; 
                
                // создать личный ключ
                return new PrivateKey(factory, null, keyOID, parameters, ecKeySpec.getS()); 
            }
            try { 
                // преобразовать тип кривой
                CurveFp curve = CurveFp.convert(ecParamSpec.getCurve());
        
                // создать параметры ключа
                IParameters parameters = new Parameters(curve, 
                    ecParamSpec.getGenerator(), ecParamSpec.getOrder()
                ); 
                // создать личный ключ
                return new PrivateKey(factory, null, keyOID, parameters, ecKeySpec.getS()); 
            }
            // при возникновении ошибки
            catch (IllegalArgumentException e) 
            { 
                // изменить тип исключения
                throw new InvalidKeySpecException(e.getMessage()); 
            }
        }
        // вызвать базовую функцию
        return super.createPrivateKey(factory, keySpec); 
    }
    // извлечь данные ключа
    @Override public KeySpec getPrivateKeySpec(
        aladdin.capi.IPrivateKey privateKey, Class<? extends KeySpec> specType) 
        throws InvalidKeySpecException, IOException
    {
        // выполнить преобразование типа
        IParameters parameters = (IParameters)privateKey.parameters();
        
        // выполнить преобразование типа
        IPrivateKey ecPrivateKey = (IPrivateKey)privateKey; 
        
        // в зависимости от типа данных
        if (specType.isAssignableFrom(ECPrivateKeySpec.class))
        {
            // вернуть данные ключа
            return new ECPrivateKeySpec(ecPrivateKey.getS(), 
                Parameters.convert(parameters)
            ); 
        }
        // вызвать базовую функцию
        return super.getPrivateKeySpec(privateKey, specType); 
    }
}
